Docker/Dockerfile

```diff
@@ -0,0 +1,31 @@
+FROM ubuntu:22.04
+
+RUN apt-get update
+RUN apt-get install wget python3 zlib1g-dev git build-essential -y
+
+# Install conda
+RUN wget https://repo.anaconda.com/miniconda/Miniconda3-py39_23.1.0-1-Linux-x86_64.sh
+RUN chmod +x ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh
+RUN ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh -b -p /opt/conda
```

Comment:
I get the following error on this step. It could be do to my Mac's architecture and I can look into it if this works for others.

```
MicrobeMod % cd Docker
Docker % docker build -t microbemod -f Dockerfile .

[+] Building 56.0s (9/22)
 => [internal] load .dockerignore                                                                                                                                                              0.0s
 => => transferring context: 2B                                                                                                                                                                0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                           0.0s
 => => transferring dockerfile: 1.14kB                                                                                                                                                         0.0s
 => [internal] load metadata for docker.io/library/ubuntu:22.04                                                                                                                                1.8s
 => [ 1/19] FROM docker.io/library/ubuntu:22.04@sha256:8eab65df33a6de2844c9aefd19efe8ddb87b7df5e9185a4ab73af936225685bb                                                                        1.9s
 => => resolve docker.io/library/ubuntu:22.04@sha256:8eab65df33a6de2844c9aefd19efe8ddb87b7df5e9185a4ab73af936225685bb                                                                          0.0s
 => => sha256:8eab65df33a6de2844c9aefd19efe8ddb87b7df5e9185a4ab73af936225685bb 1.13kB / 1.13kB                                                                                                 0.0s
 => => sha256:5b5a3f7bd48cdcd8ff523361ae281a4a645b1d7889b5a15856ea540e68fba5c4 424B / 424B                                                                                                     0.0s
 => => sha256:031631b933269c0c43a31cb8444745bf85c032d8166276b40e428843c8f54472 2.31kB / 2.31kB                                                                                                 0.0s
 => => sha256:32ba3a3141d26dce5dc8f42b11b36e32f5986433d380291ba7f98b143af3fc46 27.36MB / 27.36MB                                                                                               1.1s
 => => extracting sha256:32ba3a3141d26dce5dc8f42b11b36e32f5986433d380291ba7f98b143af3fc46                                                                                                      0.7s
 => [ 2/19] RUN apt-get update                                                                                                                                                                 6.5s
 => [ 3/19] RUN apt-get install wget python3 zlib1g-dev git build-essential -y                                                                                                                42.1s
 => [ 4/19] RUN wget https://repo.anaconda.com/miniconda/Miniconda3-py39_23.1.0-1-Linux-x86_64.sh                                                                                              2.9s
 => [ 5/19] RUN chmod +x ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh                                                                                                                            0.3s
 => ERROR [ 6/19] RUN ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh -b -p /opt/conda                                                                                                              0.3s
------
 > [ 6/19] RUN ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh -b -p /opt/conda:
#0 0.110 PREFIX=/opt/conda
#0 0.315 Unpacking payload ...
#0 0.318 qemu-x86_64: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
------
Dockerfile:9
--------------------
   7 |     RUN wget https://repo.anaconda.com/miniconda/Miniconda3-py39_23.1.0-1-Linux-x86_64.sh
   8 |     RUN chmod +x ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh
   9 | >>> RUN ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh -b -p /opt/conda
  10 |     ENV PATH=/opt/conda/bin:$PATH
  11 |     RUN conda init
--------------------
ERROR: failed to solve: process "/bin/sh -c ./Miniconda3-py39_23.1.0-1-Linux-x86_64.sh -b -p /opt/conda" did not complete successfully: exit code: 255
```

Reply:
That error is due to the architecture. You can get around it with a cross build:

```
docker buildx create --name builder --use --bootstrap
docker buildx build --load -t microbemod --platform linux/amd64 -f Dockerfile .
```

However, I get stuck at `[11/19] RUN cd meme-5.5.5 && make` with this error:
```
73.10 ../../libtool: line 1916:  1745 Segmentation fault      gcc -DHAVE_CONFIG_H -I. -I../.. -I. -I./include -w -std=gnu89 -fno-common -Wall -Wno-unused -DUNIX -D__USE_FIXED_PROTOTYPES__ -O3 -MT libxml2_la-globals.lo -MD -MP -MF .deps/libxml2_la-globals.Tpo -c globals.c -o libxml2_la-globals.o > /dev/null 2>&1
```

I think we will have to settle for a working AMD64 image.

Reply:
James instructions worked for me (Apple M2 Pro, Ventura 13.2.1). To run without a raising a platform mismatch warning:

```shell
docker run --platform linux/amd64 -v $PWD:/home/ubuntu/ -w /home/ubuntu/ microbemod -h
```

Reply:
@alexcritschristoph Is this partial result expected?

```shell
% docker run --platform linux/amd64 -v $PWD:/home/ubuntu/ -w /home/ubuntu/ microbemod call_methylation -b ./tests/test_data/test.bam -r ./tests/test_data/EcoliCVM05_GCF_000005845.2_ASM584v2_genomic.fna -t 10
23-12-20 20:09:03 INFO     Running modkit: modkit pileup -t 10 ./tests/test_data/test.bam test_low.bed -r ./tests/test_data/EcoliCVM05_GCF_000005845.2_ASM584v2_genomic.fna --only-tabs --filter-threshold 0.66
23-12-20 20:09:30 INFO     Reading low cutoff Modkit table...
23-12-20 20:09:31 INFO     Retrieving sequences
23-12-20 20:09:33 INFO     Total number of sites with any possible calls (including low quality): 116243 bases
23-12-20 20:09:33 INFO     Median coverage (stranded): 53.0x
23-12-20 20:09:33 INFO     Total number of sites with greater than min coverage (10x): 80836 bases
23-12-20 20:09:33 INFO     Potential sites (>33%) for methylation type m: 573 bases
23-12-20 20:09:33 INFO     High quality sites (>66%) for methylation type m: 421 bases
23-12-20 20:09:33 INFO     Very high quality sites (>90%) for methylation type m: 348 bases
23-12-20 20:09:33 INFO     Potential sites (>33%) for methylation type a: 1596 bases
23-12-20 20:09:33 INFO     High quality sites (>66%) for methylation type a: 845 bases
23-12-20 20:09:33 INFO     Very high quality sites (>90%) for methylation type a: 510 bases
23-12-20 20:09:33 INFO     Running for Methylation type: 6mA
23-12-20 20:09:33 INFO     Modkit table size: 57110
23-12-20 20:09:33 INFO     510 sites for STREME with cutoff 0.9 and min coverage 10
23-12-20 20:09:33 INFO     Creating background control distribution: test_a_control.fasta
23-12-20 20:09:34 INFO     Running STREME: streme . --n test_a_control.fasta -p test_a_pos.fasta -o test_a_pos_streme
23-12-20 20:10:24 INFO     Running for Methylation type: 5mC
23-12-20 20:10:24 INFO     Modkit table size: 59133
23-12-20 20:10:24 INFO     348 sites for STREME with cutoff 0.9 and min coverage 10
23-12-20 20:10:24 INFO     Creating background control distribution: test_m_control.fasta
23-12-20 20:10:26 INFO     Running STREME: streme . --n test_m_control.fasta -p test_m_pos.fasta -o test_m_pos_streme
Error: No methylated sites sent to STREME.
```

Reply:
The other program worked:

```shell
% docker run --platform linux/amd64 -v $PWD:/home/ubuntu/ -w /home/ubuntu/ microbemod annotate_rm -f ./tests/test_data/EcoliCVM05_GCF_000005845.2_ASM584v2_genomic.fna -o genome_reference -t 10
23-12-20 20:26:58 INFO     Running prodigal: prodigal -i ./tests/test_data/EcoliCVM05_GCF_000005845.2_ASM584v2_genomic.fna -a genome_reference.faa
23-12-20 20:27:36 INFO     Running HMMER: hmmsearch --cut_ga --cpu 10 --domtblout genome_reference.hits /opt/conda/lib/python3.9/site-packages/MicrobeMod/db/HMMs/RM_HMMs.hmm genome_reference.faa
23-12-20 20:28:15 INFO     Running cath: cath-resolve-hits --input-format hmmer_domtblout genome_reference.hits --hits-text-to-file genome_reference.resolved.hits
23-12-20 20:28:16 INFO     Running BLASTP against REBASE: blastp -query genome_reference.rm.genes.faa -db /opt/conda/lib/python3.9/site-packages/MicrobeMod/db/rebase_blast/all_rebase_proteins.faa -outfmt 6 -evalue 1e-5 -num_threads 10 > genome_reference.blast
```

Output:

```shell
% cat genome_reference.rm.genes.tsv
Operon  Gene    System Type     Gene type       HMM     Evalue  REBASE homolog  Homolog identity(%)     Homolog methylation     Homolog motif
RM Operon #1    NC_000913.3_1130        RM_Type_IV      RE      Type_IV_05-RM_Type_IV__Type_IV_REases   6.8e-27 SsoSE61ORF22640P        100.0           YCGR
RM Operon #2    NC_000913.3_4263        RM_Type_IV      RE      FAM_1-RM_Type_IV__Type_IV_REases        4e-149  Eco1655dMcrBCP_(Eco1655dMcrBP)  100.0
RM Operon #2    NC_000913.3_4268        RM_Type_IV      RE      FAM_0-RM_Type_IV__Type_IV_REases        2.1e-76 EcoZK126MrrP    100.0
RM Operon #2    NC_000913.3_4266        RM_Type_I       MT      Type_I_MTases_FAM_2     4.7e-252        M.SenHNK130ORF17125P    100.0   m6A     AACNNNNNNGTGC
RM Operon #2    NC_000913.3_4267        RM_Type_I       RE      Type_I_REases_FAM_2.einsi_trimmed       0.0     Msa17ORFC2P     100.0           AACNNNNNNGTGC
RM Operon #2    NC_000913.3_4265        RM_Type_I       SP      Type_I_S_52     1.8e-81 S.SenHNK130ORF17125P    100.0           AACNNNNNNGTGC
RM Operon #2    NC_000913.3_4262        RM_Type_IV      RE      FAM_2-RM_Type_IV__Type_IV_REases        8.6e-137        SenHNK130McrBCP_(SenHNK130McrCP)        100.0
Singleton #1    NC_000913.3_464 RM_Type_II      RE      Type_II_REase06 7.9e-10
Singleton #2    NC_000913.3_1932        RM_Type_II      MT      Type_II_MTases_FAM_2    1.2e-118        M.SflLIN6DcmP   100.0   m5C     CCWGG
Singleton #3    NC_000913.3_2152        RM_Type_I       RE      Type_I_REases_FAM_2.einsi_trimmed       5.1e-31
Singleton #4    NC_000913.3_3194        RM_Type_II      MT      Type_II_MTases_FAM_1    5.1e-93 M.Eco4792LORF2734P      100.0           ATGCAT
Singleton #5    NC_000913.3_3316        RM_Type_II      MT      Type_II_MTases_FAM_4    1.4e-87 M.UbaC1152DamP  100.0   m6A     GATC
```